tools/fuzzer.cpp

```diff
@@ -0,0 +1,274 @@
+  ADD_UTC_DATE,
+  ADD_BINARY,
+  ADD_EXTERNAL,
+  ADD_ILLEGAL,
+  ADD_MIN_KEY,
+  ADD_MAX_KEY,
+  ADD_MAX_VPACK_VALUE
+};
+
+struct RandomGenerator {
+  RandomGenerator() : mt64{rd()} {}
+
+  std::random_device rd;
+  std::mt19937_64 mt64;
+};
+
+std::vector<std::jthread> jthreads;
+
+
+static void finalize() {
+  for (auto& jthread : jthreads) {
+    jthread.request_stop(); // this is said to be safely called concurrently for the same object
+  }
+}
+
+
+static void usage(char* argv[]) {
+  std::cout << "Usage: " << argv[0] << " [OPTIONS] [ITERATIONS] [THREADS]"
+            << std::endl;
+  std::cout << "This program creates <iterations> random VPack or JSON structures and validates them."
+            << std::endl;
+  std::cout << "The paralelization is supplied by <threads>."
+            << std::endl;
+  std::cout << "Available options are:" << std::endl;
+  std::cout << " --vpack       create VPack."
+            << std::endl;
+  std::cout << " --json        create JSON."
+            << std::endl;
+  std::cout << " <iterations>  number of iterations. Default: 1"
+            << std::endl;
+  std::cout << " <threads>  number of threads. Default: 1"
+            << std::endl;
+}
+
+static inline bool isOption(char const* arg, char const* expected) {
+  return (strcmp(arg, expected) == 0);
+}
+
+static void addString(Builder& builder, RandomGenerator& randomGenerator) {
+  static auto &availableChars = "0123456789"
+                                "abcdefghijklmnopqrstuvwxyz"
+                                "ABCDEFGHIJKLMNOPQRSTUVWXYZ";
+  size_t length = randomGenerator.mt64() % 1000;
+  std::string s;
+  s.reserve(length);
+  while (length--) {
+    s += availableChars[randomGenerator.mt64() % (sizeof(availableChars) - 1)];
+  }
+  builder.add(Value(s));
+}
+
+static void generateVelocypack(Builder& builder, size_t depth, RandomGenerator& randomGenerator, Format const& format) {
+  while (true) {
+    RandomBuilderAdditions randomBuilderAdds = static_cast<RandomBuilderAdditions>(randomGenerator.mt64() %
+                                                                                   (format == Format::JSON
+                                                                                    ? ADD_DOUBLE
+                                                                                    : ADD_MAX_VPACK_VALUE));
+    if (depth > 10 && randomBuilderAdds <= ADD_OBJECT) {
+      continue;
+    }
+    switch (randomBuilderAdds) {
+      case ADD_ARRAY: {
+        builder.openArray(randomGenerator.mt64() % 2 ? true : false);
+        size_t numMembers = randomGenerator.mt64() % 10;
+        for (size_t i = 0; i < numMembers; ++i) {
+          generateVelocypack(builder, depth + 1, randomGenerator, format);
+        }
+        builder.close();
+        break;
+      }
+      case ADD_OBJECT: {
+        builder.openObject(randomGenerator.mt64() % 2 ? true : false);
+        size_t numMembers = randomGenerator.mt64() % 10;
```

Comment:
Magic number should be put in a constexpr somewhere outside the function (together with other magic numbers).